File: node/src/pat.ts
import {
  DEFAULT_BASE_URL,
  DEFAULT_HTTP_TIMEOUT_MS
} from './config.js';
import { decodeTokenClaims } from './auth/claims.js';
import { decodeApiError, TransportError, ValidationError } from './errors.js';
import {
  HEADER_OPERON_DID,
  HEADER_OPERON_PAYLOAD_HASH,
  buildTransactionSubmission,
  normalizeTransactionRequestAliases,
  resolvePayload,
  sanitizeOperonHeaders,
  validateTransactionRequestForSubmit
} from './transactions.js';
import {
  buildKeyId,
  canonicalSigningAlgorithm,
  type FetchFunction,
  type OperonHeaders,
  type Signature,
  type SignatureValidationResult,
  type Transaction,
  type TransactionPayload,
  type TransactionRequest,
  type Workstream,
  type WorkstreamInteractionsResponse,
  type WorkstreamParticipantsResponse
} from './types.js';

export interface ClientAPIConfig {
  baseUrl?: string;
  httpTimeoutMs?: number;
  fetchImpl?: FetchFunction;
}

export interface WorkstreamDataConfig {
  baseUrl?: string;
  httpTimeoutMs?: number;
  fetchImpl?: FetchFunction;
}

interface NormalizedConfig {
  baseUrl: string;
  httpTimeoutMs: number;
  fetchImpl: FetchFunction;
}

/**
 * Signs a payload hash using managed Operon keys with PAT authentication.
 */
export async function signHashWithPAT(
  cfg: ClientAPIConfig,
  pat: string,
  payloadHash: string,
  algorithm: string,
  options: { signal?: AbortSignal } = {}
): Promise<Signature> {
  const token = pat.trim();
  if (!token) {
    throw new ValidationError('pat is required');
  }

  const hash = payloadHash.trim();
  if (!hash) {
    throw new ValidationError('payload hash is required');
  }

  const selectedAlgorithm = canonicalSigningAlgorithm(algorithm);
  if (!selectedAlgorithm) {
    throw new ValidationError(`unsupported signing algorithm ${algorithm}`);
  }

  const normalized = normalizeConfig(cfg);
  const response = await requestJSON(normalized, {
    method: 'POST',
    path: '/v1/dids/self/sign',
    token,
    body: {
      payloadHash: hash,
      hashAlgorithm: 'SHA-256',
      algorithm: selectedAlgorithm
    },
    signal: options.signal
  });

  if (response.status >= 400) {
    throw await decodeApiError(response);
  }

  const body = (await response.json()) as { signature?: Signature };
  if (!body.signature) {
    throw new ValidationError('self sign response missing signature');
  }

  const signature: Signature = {
    algorithm: body.signature.algorithm,
    value: body.signature.value,
    keyId: body.signature.keyId
  };

  if (!signature.keyId?.trim()) {
    const claims = decodeTokenClaims(token);
    if (claims.participantDid?.trim()) {
      signature.keyId = buildKeyId(claims.participantDid);
    }
  }

  return signature;
}

/**
 * Submits a signed transaction with PAT authentication.
 */
export async function submitTransactionWithPAT(
  cfg: ClientAPIConfig,
  pat: string,
  request: TransactionRequest,
  options: { signal?: AbortSignal } = {}
): Promise<Transaction> {
  const token = pat.trim();
  if (!token) {
    throw new ValidationError('pat is required');
  }

  const req: TransactionRequest = {
    ...request,
    tags: request.tags ? [...request.tags] : undefined,
    signature: request.signature ? { ...request.signature } : undefined
  };

  normalizeTransactionRequestAliases(req);

  const claims = decodeTokenClaims(token);
  if (!req.workstreamId?.trim() && claims.workstreamId?.trim()) {
    req.workstreamId = claims.workstreamId.trim();
  }
  if (!req.sourceDid?.trim() && claims.participantDid?.trim()) {
    req.sourceDid = claims.participantDid.trim();
  }

  const { payloadHash } = resolvePayload(req.payload, req.payloadHash);
  req.payloadHash = payloadHash;

  normalizeTransactionRequestAliases(req);
  validateTransactionRequestForSubmit(req);

  const signature = {
    algorithm: req.signature!.algorithm.trim(),
    value: req.signature!.value.trim(),
    keyId: req.signature!.keyId?.trim() || undefined
  };
  const submission = buildTransactionSubmission(req, signature, payloadHash, req.timestamp ?? new Date());

  const normalized = normalizeConfig(cfg);
  const response = await requestJSON(normalized, {
    method: 'POST',
    path: '/v1/transactions',
    token,
    body: submission,
    signal: options.signal
  });

  if (response.status >= 400) {
    throw await decodeApiError(response);
  }

  const body = (await response.json()) as Record<string, unknown>;
  return deserializeTransaction(body);
}

/**
 * Validates Operon signature headers against a payload with PAT authentication.
 */
export async function validateSignatureWithPAT(
  cfg: ClientAPIConfig,
  pat: string,
  payload: TransactionPayload,
  headers: OperonHeaders,
  options: { signal?: AbortSignal } = {}
): Promise<SignatureValidationResult> {
  const token = pat.trim();
  if (!token) {
    throw new ValidationError('pat is required');
  }

  const payloadInput = resolvePayload(payload, undefined);
  if (!payloadInput.payloadBytes) {
    throw new ValidationError('payload bytes are required to validate Operon headers');
  }

  const sanitized = sanitizeOperonHeaders(headers);
  const expectedHash = sanitized[HEADER_OPERON_PAYLOAD_HASH];
  if (payloadInput.payloadHash.toLowerCase() !== expectedHash.toLowerCase()) {
    throw new ValidationError(
      `payload hash mismatch: expected ${payloadInput.payloadHash}, got ${expectedHash}`
    );
  }

  const normalized = normalizeConfig(cfg);
  const did = sanitized[HEADER_OPERON_DID];
  const response = await requestRaw(normalized, {
    method: 'POST',
    path: `/v1/dids/${encodeURIComponent(did)}/signature/verify`,
    token,
    body: payloadInput.payloadBytes,
    headers: sanitized,
    signal: options.signal
  });

  if (response.status >= 400) {
    throw await decodeApiError(response);
  }

  const body = (await response.json()) as SignatureValidationResult;
  return {
    status: body.status,
    message: body.message,
    did: body.did,
    payloadHash: body.payloadHash,
    algorithm: body.algorithm,
    keyId: body.keyId
  };
}

/**
 * Convenience wrapper around validateSignatureWithPAT for string payloads.
 */
export async function validateSignatureWithPATFromString(
  cfg: ClientAPIConfig,
  pat: string,
  payload: string,
  headers: OperonHeaders,
  options: { signal?: AbortSignal } = {}
): Promise<SignatureValidationResult> {
  return validateSignatureWithPAT(cfg, pat, payload, headers, options);
}

/**
 * Fetches workstream metadata with PAT authentication.
 */
export async function fetchWorkstream(
  cfg: WorkstreamDataConfig,
  pat: string,
  workstreamId?: string,
  options: { signal?: AbortSignal } = {}
): Promise<Workstream> {
  const token = pat.trim();
  if (!token) {
    throw new ValidationError('pat is required');
  }

  const targetWorkstream = resolveWorkstreamIdFromPAT(token, workstreamId);
  const normalized = normalizeConfig(cfg);
  const response = await requestJSON(normalized, {
    method: 'GET',
    path: `/v1/workstreams/${encodeURIComponent(targetWorkstream)}`,
    token,
    signal: options.signal
  });

  if (response.status >= 400) {
    throw await decodeApiError(response);
  }

  const body = (await response.json()) as Record<string, unknown>;
  return deserializeWorkstream(body);
}

/**
 * Fetches workstream interactions with PAT authentication.
 */
export async function fetchWorkstreamInteractions(
  cfg: WorkstreamDataConfig,
  pat: string,
  workstreamId?: string,
  options: { signal?: AbortSignal } = {}
): Promise<WorkstreamInteractionsResponse> {
  return (await fetchWorkstreamDataset(
    cfg,
    pat,
    'interactions',
    workstreamId,
    options
  )) as WorkstreamInteractionsResponse;
}

/**
 * Fetches workstream participants with PAT authentication.
 */
export async function fetchWorkstreamParticipants(
  cfg: WorkstreamDataConfig,
  pat: string,
  workstreamId?: string,
  options: { signal?: AbortSignal } = {}
): Promise<WorkstreamParticipantsResponse> {
  return (await fetchWorkstreamDataset(
    cfg,
    pat,
    'participants',
    workstreamId,
    options
  )) as WorkstreamParticipantsResponse;
}

async function fetchWorkstreamDataset(
  cfg: WorkstreamDataConfig,
  pat: string,
  resource: 'interactions' | 'participants',
  workstreamId?: string,
  options: { signal?: AbortSignal } = {}
): Promise<WorkstreamInteractionsResponse | WorkstreamParticipantsResponse> {
  const token = pat.trim();
  if (!token) {
    throw new ValidationError('pat is required');
  }

  const targetWorkstream = resolveWorkstreamIdFromPAT(token, workstreamId);
  const normalized = normalizeConfig(cfg);

  const response = await requestJSON(normalized, {
    method: 'GET',
    path: `/v1/workstreams/${encodeURIComponent(targetWorkstream)}/${resource}`,
    token,
    signal: options.signal
  });

  if (response.status >= 400) {
    throw await decodeApiError(response);
  }

  const body = (await response.json()) as {
    interactions?: Array<Record<string, unknown>>;
    participants?: Array<Record<string, unknown>>;
    totalCount?: number;
    page?: number;
    pageSize?: number;
    hasMore?: boolean;
  };

  if (resource === 'interactions') {
    return {
      interactions: (body.interactions ?? []).map((item) => ({
        id: String(item.id ?? ''),
        workstreamId: firstNonEmpty(asString(item.workstreamId), asString(item.channelId), ''),
        workstreamName: asString(item.workstreamName),
        name: asString(item.name),
        description: asString(item.description),
        status: asString(item.status),
        sourceParticipantId: asString(item.sourceParticipantId),
        targetParticipantId: asString(item.targetParticipantId),
        workstreams: asStringArray(item.workstreams),
        type: asString(item.type),
        actor: asString(item.actor),
        states: asStringArray(item.states),
        roiClassification: asString(item.roiClassification),
        roiCost: asNumber(item.roiCost),
        roiTime: asNumber(item.roiTime),
        tags: asStringArray(item.tags),
        createdAt: toDateOrUndefined(item.createdAt),
        updatedAt: toDateOrUndefined(item.updatedAt),
        version: asNumber(item.version)
      })),
      totalCount: body.totalCount ?? 0,
      page: body.page ?? 0,
      pageSize: body.pageSize ?? 0,
      hasMore: body.hasMore ?? false
    };
  }

  return {
    participants: (body.participants ?? []).map((item) => ({
      id: String(item.id ?? ''),
      did: String(item.did ?? ''),
      name: asString(item.name),
      description: asString(item.description),
      url: asString(item.url),
      status: asString(item.status),
      type: asString(item.type),
      customerId: asString(item.customerId),
      workstreamId: firstNonEmpty(asString(item.workstreamId), asString(item.channelId)),
      workstreamName: asString(item.workstreamName),
      tags: asStringArray(item.tags),
      createdAt: toDateOrUndefined(item.createdAt),
      updatedAt: toDateOrUndefined(item.updatedAt),
      version: asNumber(item.version)
    })),
    totalCount: body.totalCount ?? 0,
    page: body.page ?? 0,
    pageSize: body.pageSize ?? 0,
    hasMore: body.hasMore ?? false
  };
}

function resolveWorkstreamIdFromPAT(pat: string, override?: string): string {
  if (override?.trim()) {
    return override.trim();
  }

  const claims = decodeTokenClaims(pat);
  if (claims.workstreamId?.trim()) {
    return claims.workstreamId.trim();
  }

  throw new ValidationError(
    'workstream ID is required: token not scoped to a workstream and no override provided'
  );
}

function normalizeConfig(input: ClientAPIConfig | WorkstreamDataConfig): NormalizedConfig {
  const baseUrl = (input.baseUrl ?? DEFAULT_BASE_URL).trim();
  if (!baseUrl) {
    throw new ValidationError('baseUrl is required');
  }

  let normalizedURL: string;
  try {
    normalizedURL = new URL(baseUrl).toString().replace(/\/$/, '');
  } catch (error) {
    throw new ValidationError(
      `invalid baseUrl: ${error instanceof Error ? error.message : String(error)}`
    );
  }

  const fetchImpl =
    input.fetchImpl ??
    (typeof globalThis.fetch === 'function' ? globalThis.fetch.bind(globalThis) : undefined);
  if (!fetchImpl) {
    throw new ValidationError(
      'fetch implementation is required; supply fetchImpl when global fetch is unavailable'
    );
  }

  return {
    baseUrl: normalizedURL,
    httpTimeoutMs:
      input.httpTimeoutMs && input.httpTimeoutMs > 0 ? input.httpTimeoutMs : DEFAULT_HTTP_TIMEOUT_MS,
    fetchImpl
  };
}

interface RequestOptions {
  method: string;
  path: string;
  token: string;
  body?: unknown;
  signal?: AbortSignal;
}

async function requestJSON(cfg: NormalizedConfig, options: RequestOptions): Promise<Response> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), cfg.httpTimeoutMs);
  const url = `${cfg.baseUrl}${options.path.startsWith('/') ? options.path : `/${options.path}`}`;

  try {
    const response = await cfg.fetchImpl(url, {
      method: options.method,
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${options.token}`,
        ...(options.body !== undefined && options.body !== null
          ? { 'Content-Type': 'application/json' }
          : undefined)
      },
      body: options.body !== undefined && options.body !== null ? JSON.stringify(options.body) : undefined,
      signal: mergeSignals(controller.signal, options.signal)
    });
    return response;
  } catch (error) {
    throw new TransportError(`request ${options.method} ${url} failed`, error);
  } finally {
    clearTimeout(timeout);
  }
}

interface RawRequestOptions {
  method: string;
  path: string;
  token: string;
  body: Buffer;
  headers: Record<string, string>;
  signal?: AbortSignal;
}

async function requestRaw(cfg: NormalizedConfig, options: RawRequestOptions): Promise<Response> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), cfg.httpTimeoutMs);
  const url = `${cfg.baseUrl}${options.path.startsWith('/') ? options.path : `/${options.path}`}`;

  try {
    const response = await cfg.fetchImpl(url, {
      method: options.method,
      headers: {
        Authorization: `Bearer ${options.token}`,
        ...options.headers
      },
      body: Uint8Array.from(options.body),
      signal: mergeSignals(controller.signal, options.signal)
    });
    return response;
  } catch (error) {
    throw new TransportError(`request ${options.method} ${url} failed`, error);
  } finally {
    clearTimeout(timeout);
  }
}

function deserializeTransaction(payload: Record<string, unknown>): Transaction {
  const workstreamId = firstNonEmpty(asString(payload.workstreamId), asString(payload.channelId), '');
  return {
    id: String(payload.id ?? ''),
    correlationId: String(payload.correlationId ?? ''),
    workstreamId,
    channelId: workstreamId,
    workstreamName: asString(payload.workstreamName),
    customerId: asString(payload.customerId),
    workspaceId: asString(payload.workspaceId),
    interactionId: String(payload.interactionId ?? ''),
    timestamp: toDate(String(payload.timestamp ?? '')),
    sourceDid: String(payload.sourceDid ?? ''),
    targetDid: String(payload.targetDid ?? ''),
    state: asString(payload.state),
    stateId: asString(payload.stateId),
    stateLabel: asString(payload.stateLabel),
    roiClassification: asString(payload.roiClassification),
    roiCostIncrement: asNumber(payload.roiCostIncrement),
    roiTimeIncrement: asNumber(payload.roiTimeIncrement),
    roiCostSavings: asNumber(payload.roiCostSavings),
    roiTimeSavings: asNumber(payload.roiTimeSavings),
    roiBaseCost: asNumber(payload.roiBaseCost),
    roiBaseTime: asNumber(payload.roiBaseTime),
    roiCostSaving: asNumber(payload.roiCostSaving),
    roiTimeSaving: asNumber(payload.roiTimeSaving),
    signature: (payload.signature as Signature) ?? { algorithm: '', value: '' },
    label: asString(payload.label),
    tags: asStringArray(payload.tags),
    payloadHash: String(payload.payloadHash ?? ''),
    actorExternalId: asString(payload.actorExternalId),
    actorExternalDisplayName: asString(payload.actorExternalDisplayName),
    actorExternalSource: asString(payload.actorExternalSource),
    assigneeExternalId: asString(payload.assigneeExternalId),
    assigneeExternalDisplayName: asString(payload.assigneeExternalDisplayName),
    assigneeExternalSource: asString(payload.assigneeExternalSource),
    status: String(payload.status ?? ''),
    hcsTopicId: asString(payload.hcsTopicId),
    hcsSequenceNumber: asNumber(payload.hcsSequenceNumber),
    hcsConsensusTimestamp: asString(payload.hcsConsensusTimestamp),
    hcsTransactionId: asString(payload.hcsTransactionId),
    hcsRunningHash: asString(payload.hcsRunningHash),
    createdAt: toDate(String(payload.createdAt ?? '')),
    updatedAt: toDate(String(payload.updatedAt ?? '')),
    createdBy: asString(payload.createdBy),
    updatedBy: asString(payload.updatedBy),
    version: asNumber(payload.version)
  };
}

function deserializeWorkstream(payload: Record<string, unknown>): Workstream {
  return {
    id: String(payload.id ?? ''),
    createdAt: toDateOrUndefined(payload.createdAt),
    updatedAt: toDateOrUndefined(payload.updatedAt),
    createdBy: asString(payload.createdBy),
    updatedBy: asString(payload.updatedBy),
    version: asNumber(payload.version),
    customerId: asString(payload.customerId),
    workspaceId: asString(payload.workspaceId),
    name: asString(payload.name),
    description: asString(payload.description),
    mode: asString(payload.mode),
    type: asString(payload.type),
    status: asString(payload.status),
    states: Array.isArray(payload.states)
      ? payload.states
          .map((state) => {
            if (!state || typeof state !== 'object') {
              return undefined;
            }
            const value = state as Record<string, unknown>;
            if (!value.id || !value.name) {
              return undefined;
            }
            return {
              id: String(value.id),
              name: String(value.name),
              status: asString(value.status)
            };
          })
          .filter((entry): entry is NonNullable<typeof entry> => Boolean(entry))
      : undefined,
    defaultStateId: asString(payload.defaultStateId),
    interactionIds: asStringArray(payload.interactionIds),
    hcsTestTopicId: asString(payload.hcsTestTopicId),
    hcsLiveTopicId: asString(payload.hcsLiveTopicId)
  };
}

function toDate(value: string): Date {
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) {
    throw new ValidationError(`invalid timestamp ${value}`);
  }
  return date;
}

function toDateOrUndefined(value: unknown): Date | undefined {
  if (typeof value !== 'string' || !value.trim()) {
    return undefined;
  }
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? undefined : date;
}

function asString(value: unknown): string | undefined {
  if (typeof value !== 'string') {
    return undefined;
  }
  const trimmed = value.trim();
  return trimmed || undefined;
}

function asNumber(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined;
}

function asStringArray(value: unknown): string[] | undefined {
  if (!Array.isArray(value)) {
    return undefined;
  }
  const items = value
    .map((entry) => (typeof entry === 'string' ? entry.trim() : ''))
    .filter((entry) => entry.length > 0);
  return items.length > 0 ? items : undefined;
}

function firstNonEmpty(...values: Array<string | undefined>): string {
  for (const value of values) {
    if (value?.trim()) {
      return value.trim();
    }
  }
  return '';
}

function mergeSignals(primary: AbortSignal, secondary?: AbortSignal): AbortSignal {
  if (!secondary) {
    return primary;
  }
  if (secondary.aborted) {
    return secondary;
  }
  const controller = new AbortController();
  const abort = () => controller.abort();
  primary.addEventListener('abort', abort, { once: true });
  secondary.addEventListener('abort', abort, { once: true });
  return controller.signal;
}

File: python/operon_sdk/pat.py
from __future__ import annotations

"""PAT-scoped helper functions for Operon client APIs."""

from dataclasses import dataclass
from datetime import datetime, timezone
from hashlib import sha256
from typing import Any, Dict, Optional
from urllib.parse import quote

import httpx

from .auth.token_provider import decode_token_claims
from .client import (
    HEADER_OPERON_DID,
    HEADER_OPERON_PAYLOAD_HASH,
    HEADER_OPERON_SIGNATURE,
    HEADER_OPERON_SIGNATURE_ALGO,
    HEADER_OPERON_SIGNATURE_KEY,
    OperonHeaders,
    _base64url_encode,
    _sanitize_operon_headers,
)
from .config import DEFAULT_BASE_URL, DEFAULT_HTTP_TIMEOUT
from .errors import ApiError, TransportError, ValidationError
from .models import (
    Signature,
    SignatureValidationResult,
    Transaction,
    TransactionRequest,
    Workstream,
    WorkstreamInteractionsResponse,
    WorkstreamParticipantsResponse,
    build_key_id,
    canonical_signing_algorithm,
    decode_payload_base64,
    validate_payload_hash_format,
)


@dataclass(slots=True)
class ClientAPIConfig:
    """Base configuration required for PAT-scoped client API calls."""

    base_url: str = DEFAULT_BASE_URL
    http_timeout: float = DEFAULT_HTTP_TIMEOUT
    http_client: Optional[httpx.AsyncClient] = None


@dataclass(slots=True)
class WorkstreamDataConfig:
    """Base configuration required for PAT-scoped workstream catalogue calls."""

    base_url: str = DEFAULT_BASE_URL
    http_timeout: float = DEFAULT_HTTP_TIMEOUT
    http_client: Optional[httpx.AsyncClient] = None


async def sign_hash_with_pat(
    cfg: ClientAPIConfig,
    pat: str,
    payload_hash: str,
    algorithm: str,
) -> Signature:
    """Sign a payload hash using a PAT and managed keys."""

    pat = pat.strip()
    if not pat:
        raise ValidationError("pat is required")

    payload_hash = payload_hash.strip()
    if not payload_hash:
        raise ValidationError("payload hash is required")
    try:
        validate_payload_hash_format(payload_hash)
    except ValueError as exc:
        raise ValidationError(str(exc)) from exc

    canonical = canonical_signing_algorithm(algorithm)
    if not canonical:
        raise ValidationError(f"unsupported signing algorithm {algorithm}")

    base_url, client, owns = _normalize_client_config(cfg)
    try:
        response = await _request_json(
            client,
            "POST",
            f"{base_url}/v1/dids/self/sign",
            token=pat,
            payload={
                "payloadHash": payload_hash,
                "hashAlgorithm": "SHA-256",
                "algorithm": canonical,
            },
        )

        payload = response.json()
        if not isinstance(payload, dict) or "signature" not in payload:
            raise ValidationError("self sign response missing signature")

        signature = Signature.model_validate(payload["signature"])
        if not signature.key_id.strip():
            claims = decode_token_claims(pat)
            if claims.participant_did:
                signature.key_id = build_key_id(claims.participant_did)
        return signature
    finally:
        if owns:
            await client.aclose()


async def submit_transaction_with_pat(
    cfg: ClientAPIConfig,
    pat: str,
    request: TransactionRequest,
) -> Transaction:
    """Submit a signed transaction using a PAT."""

    if not isinstance(request, TransactionRequest):
        raise ValidationError("request must be TransactionRequest")

    pat = pat.strip()
    if not pat:
        raise ValidationError("pat is required")

    claims = decode_token_claims(pat)
    if not request.workstream_id.strip() and claims.workstream_id:
        request.workstream_id = claims.workstream_id
    if not request.source_did.strip() and claims.participant_did:
        request.source_did = claims.participant_did

    _, payload_hash = request.resolve_payload()
    request.payload_hash = payload_hash

    try:
        request.validate_for_submit()
    except ValueError as exc:
        raise ValidationError(str(exc)) from exc

    timestamp = request.timestamp or datetime.now(timezone.utc)
    submission: Dict[str, Any] = {
        "correlationId": request.correlation_id,
        "workstreamId": request.workstream_id,
        "interactionId": request.interaction_id,
        "timestamp": timestamp.isoformat(),
        "sourceDid": request.source_did,
        "targetDid": request.target_did,
        "roiClassification": request.roi_classification,
        "roiCost": request.roi_cost,
        "roiTime": request.roi_time,
        "state": request.state,
        "stateId": request.state_id,
        "stateLabel": request.state_label,
        "roiBaseCost": request.roi_base_cost,
        "roiBaseTime": request.roi_base_time,
        "roiCostSaving": request.roi_cost_saving,
        "roiTimeSaving": request.roi_time_saving,
        "signature": request.signature.model_dump(by_alias=True),
        "payloadHash": payload_hash,
        "actorExternalId": request.actor_external_id,
        "actorExternalDisplayName": request.actor_external_display_name,
        "actorExternalSource": request.actor_external_source,
        "assigneeExternalId": request.assignee_external_id,
        "assigneeExternalDisplayName": request.assignee_external_display_name,
        "assigneeExternalSource": request.assignee_external_source,
        "customerId": request.customer_id,
        "workspaceId": request.workspace_id,
        "createdBy": request.created_by,
    }

    if request.label.strip():
        submission["label"] = request.label.strip()

    if request.tags:
        cleaned_tags = [tag.strip() for tag in request.tags if tag.strip()]
        if cleaned_tags:
            submission["tags"] = cleaned_tags

    base_url, client, owns = _normalize_client_config(cfg)
    try:
        response = await _request_json(
            client,
            "POST",
            f"{base_url}/v1/transactions",
            token=pat,
            payload=submission,
        )
        return Transaction.model_validate(response.json())
    finally:
        if owns:
            await client.aclose()


async def validate_signature_with_pat(
    cfg: ClientAPIConfig,
    pat: str,
    payload: bytes,
    headers: OperonHeaders,
) -> SignatureValidationResult:
    """Validate Operon signature headers against payload bytes using PAT auth."""

    pat = pat.strip()
    if not pat:
        raise ValidationError("pat is required")

    sanitized = _sanitize_operon_headers(headers)

    digest = sha256(payload).digest()
    computed_hash = _base64url_encode(digest)
    expected_hash = sanitized[HEADER_OPERON_PAYLOAD_HASH]
    if computed_hash.lower() != expected_hash.lower():
        raise ValidationError(
            f"payload hash mismatch: expected {computed_hash}, got {expected_hash}"
        )

    did = sanitized[HEADER_OPERON_DID]

    base_url, client, owns = _normalize_client_config(cfg)
    try:
        request_headers = {"Authorization": f"Bearer {pat}"}
        request_headers.update(sanitized)
        try:
            response = await client.post(
                f"{base_url}/v1/dids/{quote(did, safe='')}/signature/verify",
                content=payload,
                headers=request_headers,
            )
        except httpx.HTTPError as exc:
            raise TransportError("perform signature validation request", original=exc) from exc

        if response.status_code >= 400:
            raise ApiError.from_response(response)

        return SignatureValidationResult.model_validate(response.json())
    finally:
        if owns:
            await client.aclose()


async def validate_signature_with_pat_from_string(
    cfg: ClientAPIConfig,
    pat: str,
    payload: str,
    headers: OperonHeaders,
) -> SignatureValidationResult:
    """Convenience wrapper around validate_signature_with_pat for text payloads."""

    return await validate_signature_with_pat(cfg, pat, payload.encode(), headers)


async def fetch_workstream(
    cfg: WorkstreamDataConfig,
    pat: str,
    workstream_id: Optional[str] = None,
) -> Workstream:
    """Fetch workstream details using a PAT."""

    base_url, client, owns = _normalize_workstream_config(cfg)
    target_workstream = _resolve_workstream_id_from_pat(pat, workstream_id)

    try:
        response = await _request_json(
            client,
            "GET",
            f"{base_url}/v1/workstreams/{quote(target_workstream, safe='')}",
            token=pat,
        )
        return Workstream.model_validate(response.json())
    finally:
        if owns:
            await client.aclose()


async def fetch_workstream_interactions(
    cfg: WorkstreamDataConfig,
    pat: str,
    workstream_id: Optional[str] = None,
) -> WorkstreamInteractionsResponse:
    """Fetch workstream interactions using a PAT."""

    base_url, client, owns = _normalize_workstream_config(cfg)
    target_workstream = _resolve_workstream_id_from_pat(pat, workstream_id)

    try:
        response = await _request_json(
            client,
            "GET",
            f"{base_url}/v1/workstreams/{quote(target_workstream, safe='')}/interactions",
            token=pat,
        )
        return WorkstreamInteractionsResponse.model_validate(response.json())
    finally:
        if owns:
            await client.aclose()


async def fetch_workstream_participants(
    cfg: WorkstreamDataConfig,
    pat: str,
    workstream_id: Optional[str] = None,
) -> WorkstreamParticipantsResponse:
    """Fetch workstream participants using a PAT."""

    base_url, client, owns = _normalize_workstream_config(cfg)
    target_workstream = _resolve_workstream_id_from_pat(pat, workstream_id)

    try:
        response = await _request_json(
            client,
            "GET",
            f"{base_url}/v1/workstreams/{quote(target_workstream, safe='')}/participants",
            token=pat,
        )
        return WorkstreamParticipantsResponse.model_validate(response.json())
    finally:
        if owns:
            await client.aclose()


def _resolve_workstream_id_from_pat(pat: str, override: Optional[str] = None) -> str:
    pat = pat.strip()
    if not pat:
        raise ValidationError("pat is required")

    if override and override.strip():
        return override.strip()

    claims = decode_token_claims(pat)
    if claims.workstream_id:
        return claims.workstream_id

    raise ValidationError(
        "workstream ID is required: token not scoped to a workstream and no override provided"
    )


async def _request_json(
    client: httpx.AsyncClient,
    method: str,
    url: str,
    *,
    token: str,
    payload: Optional[Dict[str, Any]] = None,
) -> httpx.Response:
    headers = {"Accept": "application/json", "Authorization": f"Bearer {token}"}

    try:
        response = await client.request(method, url, json=payload, headers=headers)
    except httpx.HTTPError as exc:
        raise TransportError(f"{method} {url}", original=exc) from exc

    if response.status_code >= 400:
        raise ApiError.from_response(response)

    return response


def _normalize_client_config(cfg: ClientAPIConfig) -> tuple[str, httpx.AsyncClient, bool]:
    base_url = (cfg.base_url or DEFAULT_BASE_URL).strip().rstrip("/")
    if not base_url.startswith("http://") and not base_url.startswith("https://"):
        raise ValidationError(f"invalid base URL: {cfg.base_url}")

    if cfg.http_client:
        return base_url, cfg.http_client, False

    return base_url, httpx.AsyncClient(timeout=cfg.http_timeout), True


def _normalize_workstream_config(cfg: WorkstreamDataConfig) -> tuple[str, httpx.AsyncClient, bool]:
    base_url = (cfg.base_url or DEFAULT_BASE_URL).strip().rstrip("/")
    if not base_url.startswith("http://") and not base_url.startswith("https://"):
        raise ValidationError(f"invalid base URL: {cfg.base_url}")

    if cfg.http_client:
        return base_url, cfg.http_client, False

    return base_url, httpx.AsyncClient(timeout=cfg.http_timeout), True


__all__ = [
    "ClientAPIConfig",
    "WorkstreamDataConfig",
    "decode_payload_base64",
    "fetch_workstream",
    "fetch_workstream_interactions",
    "fetch_workstream_participants",
    "sign_hash_with_pat",
    "submit_transaction_with_pat",
    "validate_signature_with_pat",
    "validate_signature_with_pat_from_string",
]
